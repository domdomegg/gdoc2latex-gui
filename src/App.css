body {
    margin: 0;
    padding: 0;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    color: white;
    background-color: #0B3954;
  }
  
  header {
    color: #0B3954;
    background-color: white;
  }
  
  header * {
    margin: 0;
  }
  
  header, section {
    padding: 36px;
    line-height: 1.5;
  }
  
  header div, section {
    max-width: 750px;
    margin: 0 auto;
  }
  
  section {
    text-align: center;
  }
  
  code, textarea {
    font-family: source-code-pro, 'Source Code Pro', Menlo, Monaco, Consolas, Fira Code, 'Courier New', monospace;
  }
  
  textarea {
    display: block;
    width: 80%;
    margin: 0 auto;
    padding: 16px;
    border-radius: 4px;
    resize: none;
  }
  
  button {
    display: block;
    margin: 12px auto;
    color: white;
    background-color: #156FA2;
    border: none;
    border-radius: 4px;
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;
    padding: 12px;
    font-size: 125%;
    cursor: pointer;
    transition: 0.15s background-color ease;
  }
  
  button:hover {
    background-color: #177BB5;
  }
  
  button:active {
    background-color: #10567F;
  }
  
  button:disabled {
    cursor: not-allowed;
    background-color: #5b5c5c;
    color: #aaa;
  }
  
  .linkButton {
    padding: 0;
    margin: 0;
    font-size: 100%;
    display: inline;
    background-color: transparent;
    color: #b0e2ff;
  }
  
  .linkButton:hover, .linkButton:active {
    background-color: transparent;
  }
  
  /* .highlight {
    white-space: nowrap;
    padding: 2px 4px;
    border: 1px solid #ccc;
    border-radius: 4px;
    box-shadow: 0 0.125rem 0.25rem rgb(0 0 0 / 30%);
  } */