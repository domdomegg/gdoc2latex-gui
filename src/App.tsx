import './App.css';
import gdoc2latex from 'gdoc2latex';
import React, { useEffect, useState } from 'react';
import { exampleGdoc } from './exampleGdoc';
import DragDropFile from './DragDropFile';
import { unzipSync } from 'fflate';

const App = () => {
  const [gdoc, setGdoc] = useState('');
  const [display, setDisplay] = useState<"input" | "output">('input');

  return (
    <div>
      <header>
        <div>
          <h1>gdoc2latex GUI</h1>
          <p>Convert a Google Doc to LaTeX using <a href="https://github.com/domdomegg/gdoc2latex">gdoc2latex</a></p>
        </div>
      </header>
      {display === "input"
        ? <InputView setDisplay={setDisplay} gdoc={gdoc} setGdoc={setGdoc} />
        : <OutputView setDisplay={setDisplay} gdoc={gdoc} />}
    </div>
  );
}

const InputView = ({ setDisplay, gdoc, setGdoc }: { setDisplay: (display: "input" | "output") => void, gdoc: string, setGdoc: (str: string) => void }) => {
  const handleFiles = async (fileList: FileList) => {
    try {
      const html = await getHtmlFromFileList(fileList);
      setGdoc(html);
      setDisplay('output');
    } catch (err) {
      alert('Error: ' + (err instanceof Error ? err.message : String(err)))
    }
  }

  return (
    <section>
      <DragDropFile handleFiles={handleFiles} />
      <h2>Drag or paste in an HTML copy of your Google Doc</h2>
      <p>
        Get this from Google Docs with <span className="highlight">File ➜ Download ➜ Web page (.html)</span><br/>
        <button className="linkButton" onClick={() => { setGdoc(exampleGdoc); setDisplay('output') }}>Or see an example</button>.
      </p>
      <input id="file_input" type="file" accept="text/html, application/zip" onChange={e => e.target.files && handleFiles(e.target.files)} />
      <button onClick={() => document.getElementById('file_input')?.click()} style={{ marginTop: "40px" }}>Upload file</button>
      <div className='or-bar'>
        <div></div>
        <p>OR</p>
        <div></div>
      </div>
      <textarea
        autoFocus
        value={gdoc}
        rows={10}
        onChange={e => setGdoc(e.target.value)}
        onKeyDown={e => { if (e.key === "Enter") { setDisplay('output') } }}
        placeholder="<html><head><meta content=&quot;text/html; charset=UTF-8&quot; http-equiv=&quot;content-type&quot;>..."
      />
      {gdoc && !gdoc.startsWith('<html>') && <p>This doesn't start with <code>&lt;html&gt;</code>. Sure you have pasted in the HTML, not the doc itself?</p>}
      {gdoc && <button onClick={() => setDisplay('output')}>Gimme! Gimme! Gimme! (A LaTeX file)</button>}
    </section>
  );
}

const OutputView = ({ setDisplay, gdoc }: { setDisplay: (display: "input" | "output") => void, gdoc: string }) => {
  const response = catchAndTransformError(() => gdoc2latex({ inputHTML: gdoc }));

  useEffect(() => {
    const handler = (event: KeyboardEvent) => {
      if (event.key === "Escape") {
        setDisplay('input');
      }
    }
  
    document.addEventListener('keydown', handler)
    return () => {
      document.removeEventListener('keydown', handler)
    }
  }, [setDisplay]);

  const download = (filename: string, content: string) => {
    const element = document.createElement('a');
    element.setAttribute('href', 'data:text/plain;charset=utf-8,' + encodeURIComponent(content));
    element.setAttribute('download', filename);
    element.style.display = 'none';
    document.body.appendChild(element);
    element.click();
    document.body.removeChild(element);
  }

  if (response.accepted) {
    return (
      <section>
        <h2>Here's your LaTeX:</h2>
        <textarea value={response.result.latex} rows={10} readOnly></textarea>
        {response.result.bibtex && <>
          <h2>and your BibTeX:</h2>
          <textarea value={response.result.bibtex} rows={10} readOnly></textarea>
        </>}
        <button onClick={() => {
          download('index.tex', response.result.latex);
          if (response.result.bibtex) download('index.bib', response.result.bibtex);
        }}>Download</button>
        <button className="linkButton" onClick={() => setDisplay('input')}>Go back</button>
      </section>
    );
  }

  return (
    <section>
      <h2>Something went wrong</h2>
      <p>Make sure that you uploaded a HTML doc, pasted the entire thing in,<br/>and formatted your doc as per the gdoc2latex instructions.</p>
      <p>In case it helps, here's the internal error we got:</p>
      <textarea value={response.message} rows={10} readOnly></textarea>
      <p>We've also logged the error to the console, where you can see a full stacktrace.</p>
      <br/>
      <button className="linkButton" onClick={() => setDisplay('input')}>Go back</button>
    </section>
  )
}

type Response<T> = Accepted<T> | Rejected;
interface Accepted<T> { accepted: true, result: T };
interface Rejected { accepted: false, message: string };

function catchAndTransformError<T>(fn: () => T): Response<T> {
  try {
    return { accepted: true, result: fn() };
  } catch (err) {
    console.error(err);
    return {
      accepted: false,
      message: err instanceof Error
        ? 'Error: ' + err.message
        : 'An error occurred converting that document'
    };
  }
}

const getHtmlFromFileList = async (fileList: FileList): Promise<string> => {
  if (fileList.length !== 1) {
    throw new Error('Please select exactly one file.');
  }

  const file = fileList[0];
  const fileName = file.name.toLowerCase();

  if (file.type !== "application/zip" && file.type !== "text/html") {
    throw new Error('Please select either a ZIP or HTML file.');
  }

  if (file.type === "application/zip") {
    const arrayBuffer = await file.arrayBuffer();
    const uint8Array = new Uint8Array(arrayBuffer);
    const unzipped: { [path: string]: Uint8Array } = unzipSync(uint8Array);

    const htmlFiles = Object.keys(unzipped).filter(path => path.toLowerCase().endsWith('.html'));

    if (htmlFiles.length === 0) {
      throw new Error('No HTML file found in the ZIP archive.');
    }

    if (htmlFiles.length > 1) {
      throw new Error('Multiple HTML files found in the ZIP archive.');
    }

    const htmlContent = new TextDecoder().decode(unzipped[htmlFiles[0]]);
    return htmlContent;
  }

  // Process HTML file
  if (fileName.endsWith('.html')) {
    return await file.text();
  }

  // This should never be reached due to earlier checks, but TypeScript might complain without it
  throw new Error('Unexpected file type.');
}

export default App;
