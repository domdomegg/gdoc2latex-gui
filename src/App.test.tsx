import { fireEvent, render, screen } from '@testing-library/react';
import App from './App';
import { exampleGdoc } from './exampleGdoc';

const exampleGdocSimple = `<html><head><meta content="text/html; charset=UTF-8" http-equiv="content-type"><style type="text/css">ol{margin:0;padding:0}table td,table th{padding:0}.c0{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:11pt;font-family:"Arial";font-style:normal}.c3{padding-top:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:left;height:11pt}.c1{color:#000000;font-weight:400;text-decoration:none;vertical-align:baseline;font-size:26pt;font-family:"Arial";font-style:normal}.c2{padding-top:0pt;padding-bottom:3pt;line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.c5{padding-top:0pt;padding-bottom:0pt;line-height:1.15;orphans:2;widows:2;text-align:left}.c4{background-color:#ffffff;max-width:451.3pt;padding:72pt 72pt 72pt 72pt}.title{padding-top:0pt;color:#000000;font-size:26pt;padding-bottom:3pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}.subtitle{padding-top:0pt;color:#666666;font-size:15pt;padding-bottom:16pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}li{color:#000000;font-size:11pt;font-family:"Arial"}p{margin:0;color:#000000;font-size:11pt;font-family:"Arial"}h1{padding-top:20pt;color:#000000;font-size:20pt;padding-bottom:6pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h2{padding-top:18pt;color:#000000;font-size:16pt;padding-bottom:6pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h3{padding-top:16pt;color:#434343;font-size:14pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h4{padding-top:14pt;color:#666666;font-size:12pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h5{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;orphans:2;widows:2;text-align:left}h6{padding-top:12pt;color:#666666;font-size:11pt;padding-bottom:4pt;font-family:"Arial";line-height:1.15;page-break-after:avoid;font-style:italic;orphans:2;widows:2;text-align:left}</style></head><body class="c4"><p class="c2 title" id="h.mkd460etze8q"><span class="c1">Simple doc</span></p><p class="c3"><span class="c0"></span></p><p class="c5"><span class="c0">I am a very simple document</span></p></body></html>`;

test('renders gdoc2latex GUI title', () => {
  render(<App />);
  const title = screen.queryByText(/gdoc2latex GUI/i);
  expect(title).toBeInTheDocument();
});

test('gives LaTeX and BibTeX output when clicking example', () => {
  render(<App />);
  expect(screen.queryByText(/Here's your LaTeX/)).not.toBeInTheDocument();
  expect(screen.queryByText(/and your BibTeX/)).not.toBeInTheDocument();
  
  fireEvent.click(screen.getByText(/example/));
  
  expect(screen.queryByText(/Here's your LaTeX/)).toBeInTheDocument();
  expect(screen.queryByText(/and your BibTeX/)).toBeInTheDocument();
  expect(screen.queryByText(/Here’s some content, originally written in Google Docs. It demonstrates using some of the features of gdoc2latex./)).toBeInTheDocument();
  expect(screen.queryByText(/author={Alexander Feder}/)).toBeInTheDocument();
});

test('gives LaTeX and BibTeX output when entering demo example', () => {
  const renderResult = render(<App />);
  expect(screen.queryByText(/Here's your LaTeX/)).not.toBeInTheDocument();  
  
  fireEvent.change(renderResult.container.querySelector('textarea')!, { target: { value: exampleGdoc } });
  fireEvent.click(screen.getByText(/Gimme! Gimme! Gimme!/));

  expect(screen.queryByText(/Here's your LaTeX/)).toBeInTheDocument();
  expect(screen.queryByText(/Here’s some content, originally written in Google Docs. It demonstrates using some of the features of gdoc2latex./)).toBeInTheDocument();
  expect(screen.queryByText(/and your BibTeX/)).toBeInTheDocument();
  expect(screen.queryByText(/author={Alexander Feder}/)).toBeInTheDocument();
});

test('gives LaTeX and BibTeX output when entering simple example', () => {
  const renderResult = render(<App />);
  expect(screen.queryByText(/Here's your LaTeX/)).not.toBeInTheDocument();  
  
  fireEvent.change(renderResult.container.querySelector('textarea')!, { target: { value: exampleGdocSimple } });
  fireEvent.click(screen.getByText(/Gimme! Gimme! Gimme!/));
  
  expect(screen.queryByText(/Here's your LaTeX/)).toBeInTheDocument();
  expect(screen.queryByText(/I am a very simple document/)).toBeInTheDocument();
  expect(screen.queryByText(/and your BibTeX/)).not.toBeInTheDocument();
});

test('can click example and go back with button', () => {
  render(<App />);
  expect(screen.queryByText(/Here's your LaTeX/)).not.toBeInTheDocument();  
  
  fireEvent.click(screen.getByText(/example/));
  
  expect(screen.queryByText(/Here's your LaTeX/)).toBeInTheDocument();

  fireEvent.click(screen.getByText(/Go back/));

  expect(screen.queryByText(/Here's your LaTeX/)).not.toBeInTheDocument();  
});

test('can click example and go back with esc key', () => {
  const renderResult = render(<App />);
  expect(screen.queryByText(/Here's your LaTeX/)).not.toBeInTheDocument();  
  
  fireEvent.click(screen.getByText(/example/));
  
  expect(screen.queryByText(/Here's your LaTeX/)).toBeInTheDocument();

  fireEvent.keyDown(renderResult.container, { key: 'Escape' });

  expect(screen.queryByText(/Here's your LaTeX/)).not.toBeInTheDocument();  
});

test('can enter example and go back with button', () => {
  const renderResult = render(<App />);
  expect(screen.queryByText(/Here's your LaTeX/)).not.toBeInTheDocument();  
  
  fireEvent.change(renderResult.container.querySelector('textarea')!, { target: { value: exampleGdoc } });
  fireEvent.click(screen.getByText(/Gimme! Gimme! Gimme!/));

  expect(screen.queryByText(/Here's your LaTeX/)).toBeInTheDocument();

  fireEvent.click(screen.getByText(/Go back/));

  expect(screen.queryByText(/Here's your LaTeX/)).not.toBeInTheDocument(); 
});

test('can enter invalid example and go back with button', () => {
  const renderResult = render(<App />);
  expect(screen.queryByText(/Something went wrong/)).not.toBeInTheDocument();
  expect(screen.queryByText(/Here's your LaTeX/)).not.toBeInTheDocument();  
  
  fireEvent.change(renderResult.container.querySelector('textarea')!, { target: { value: 'Not valid' } });
  fireEvent.click(screen.getByText(/Gimme! Gimme! Gimme!/));

  expect(screen.queryByText(/Something went wrong/)).toBeInTheDocument();
  expect(screen.queryByText(/Here's your LaTeX/)).not.toBeInTheDocument(); 
  
  fireEvent.click(screen.getByText(/Go back/));
  
  expect(screen.queryByText(/Something went wrong/)).not.toBeInTheDocument();
  expect(screen.queryByText(/Here's your LaTeX/)).not.toBeInTheDocument(); 
});

test('user journey matches snapshots', () => {
  const renderResult = render(<App />);

  // Homepage
  expect(renderResult.container).toMatchSnapshot(); 

  // Error
  fireEvent.change(renderResult.container.querySelector('textarea')!, { target: { value: 'Not valid' } });
  fireEvent.click(screen.getByText(/Gimme! Gimme! Gimme!/));
  expect(renderResult.container).toMatchSnapshot();

  // Success
  fireEvent.click(screen.getByText(/Go back/));
  fireEvent.change(renderResult.container.querySelector('textarea')!, { target: { value: exampleGdocSimple } });
  fireEvent.click(screen.getByText(/Gimme! Gimme! Gimme!/));
  expect(renderResult.container).toMatchSnapshot();
})