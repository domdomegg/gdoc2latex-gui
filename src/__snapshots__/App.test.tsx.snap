// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`user journey matches snapshots 1`] = `
<div>
  <div>
    <header>
      <div>
        <h1>
          gdoc2latex GUI
        </h1>
        <p>
          Convert a Google Doc to LaTeX using 
          <a
            href="https://github.com/domdomegg/gdoc2latex"
          >
            gdoc2latex
          </a>
        </p>
      </div>
    </header>
    <section>
      <h2>
        Paste in an HTML copy of your Google Doc
      </h2>
      <p>
        Get this from Google Docs with 
        <span
          class="highlight"
        >
          File ➜ Download ➜ Web page (.html)
        </span>
        <br />
        (you may need to unzip it first). 
        <button
          class="linkButton"
        >
          Or see an example
        </button>
        .
      </p>
      <textarea
        placeholder="<html><head><meta content=\\"text/html; charset=UTF-8\\" http-equiv=\\"content-type\\">..."
        rows="10"
      />
      
      <button
        disabled=""
      >
        Gimme! Gimme! Gimme! (A LaTeX file)
      </button>
    </section>
  </div>
</div>
`;

exports[`user journey matches snapshots 2`] = `
<div>
  <div>
    <header>
      <div>
        <h1>
          gdoc2latex GUI
        </h1>
        <p>
          Convert a Google Doc to LaTeX using 
          <a
            href="https://github.com/domdomegg/gdoc2latex"
          >
            gdoc2latex
          </a>
        </p>
      </div>
    </header>
    <section>
      <h2>
        Something went wrong
      </h2>
      <p>
        Make sure that you uploaded a HTML doc, pasted the entire thing in,
        <br />
        and formatted your doc as per the gdoc2latex instructions.
      </p>
      <p>
        In case it helps, here's the internal error we got:
      </p>
      <textarea
        readonly=""
        rows="10"
      >
        Error: Cannot read properties of undefined (reading '0')
      </textarea>
      <br />
      <button
        class="linkButton"
      >
        Go back
      </button>
    </section>
  </div>
</div>
`;

exports[`user journey matches snapshots 3`] = `
<div>
  <div>
    <header>
      <div>
        <h1>
          gdoc2latex GUI
        </h1>
        <p>
          Convert a Google Doc to LaTeX using 
          <a
            href="https://github.com/domdomegg/gdoc2latex"
          >
            gdoc2latex
          </a>
        </p>
      </div>
    </header>
    <section>
      <h2>
        Here's your LaTeX:
      </h2>
      <textarea
        readonly=""
        rows="10"
      >
        \\title{\\textbf{Simple doc}}
\\documentclass[12pt]{article}

\\usepackage[a4paper,left=2.54cm,right=2.54cm,top=2.54cm,bottom=2.54cm]{geometry}
\\usepackage[T1]{fontenc}
\\usepackage[utf8]{inputenc}
\\usepackage{lmodern}
\\usepackage{parskip}
\\usepackage{hyperref}
\\usepackage{graphicx}
\\usepackage{adjustbox}
\\usepackage{natbib}
\\usepackage[dvipsnames]{xcolor}
\\usepackage{minted}

\\setlength\\parindent{0em}

\\bibliographystyle{abbrvnat}
\\setcitestyle{super,citesep={,},open={},close={}}

\\definecolor{lightgrey}{rgb}{0.95,0.95,0.95}
\\setminted{bgcolor=lightgrey}
\\setmintedinline{bgcolor=lightgrey,escapeinside=||,mathescape=true}

\\renewcommand{\\thefootnote}{\\roman{footnote}}

\\usepackage{titling}
\\predate{}
\\date{}
\\postdate{}
\\preauthor{}
\\author{}
\\postauthor{}

\\setcounter{tocdepth}{1}

\\begin{document}

\\maketitle


I am a very simple document

\\bibliography{index}

\\end{document}
      </textarea>
      <button>
        Download
      </button>
      <button
        class="linkButton"
      >
        Go back
      </button>
    </section>
  </div>
</div>
`;
